File: src/main/java/com/joraph/debug/JoraphDebug.java
package com.joraph.debug;

import java.util.Set;

import com.joraph.ObjectGraph;
import com.joraph.plan.ExecutionPlan;

public class JoraphDebug {

	private static ThreadLocal<DebugInfo> debugInfo = new ThreadLocal<>();

	public static void setThreadDebugInfo(DebugInfo info) {
		debugInfo.set(info);
	}

	public static void clearThreadDebugInfo() {
		debugInfo.set(null);
	}

	public static void startDebug() {
		debugInfo.set(new DebugInfo());
	}

	public static DebugInfo finishDebug() {
		DebugInfo info = getDebugInfo();
		debugInfo.set(null);
		return info;
	}

	public static DebugInfo getDebugInfo() {
		return debugInfo.get();
	}

	public static boolean hasDebugInfo() {
		return debugInfo.get()!=null;
	}

	public static void addObjectGraph(ObjectGraph objectGraph) {
		if (!hasDebugInfo()) {
			return;
		}
		getDebugInfo().addObjectGraph(objectGraph);
	}

	public static void addExecutionPlan(ExecutionPlan executionPlan) {
		if (!hasDebugInfo()) {
			return;
		}
		getDebugInfo().addExecutionPlan(executionPlan);
	}

	public static void addLoaderDebug(LoaderDebug loaderDebug) {
		if (!hasDebugInfo()) {
			return;
		}
		getDebugInfo().addLoaderDebug(loaderDebug);
	}

	public static void addLoaderDebug(
			Class<?> entityClass, Long loaderTimeMillis, Integer loadedEntityCount, Set<Object> ids) {
		if (!hasDebugInfo()) {
			return;
		}
		getDebugInfo().addLoaderDebug(
				entityClass, loaderTimeMillis,
				loadedEntityCount, ids);
	}

}

File: src/main/kotlin/com/joraph/debug/JoraphDebug.kt
package com.joraph.debug

import com.joraph.ObjectGraph

object JoraphDebug {
    private val debugInfo = ThreadLocal<DebugInfo?>()

    @JvmStatic
    fun setThreadDebugInfo(info: DebugInfo?) {
        debugInfo.set(info)
    }

    @JvmStatic
    fun clearThreadDebugInfo() {
        debugInfo.set(null)
    }

    @JvmStatic
    fun startDebug() {
        debugInfo.set(DebugInfo())
    }

    @JvmStatic
    fun finishDebug(): DebugInfo? {
        val info = getDebugInfo()
        debugInfo.set(null)
        return info
    }

    @JvmStatic
    fun getDebugInfo(): DebugInfo? {
        return debugInfo.get()
    }

    @JvmStatic
    fun hasDebugInfo(): Boolean {
        return debugInfo.get() != null
    }

    @JvmStatic
    fun addObjectGraph(objectGraph: ObjectGraph) {
        if (!hasDebugInfo()) {
            return
        }
        getDebugInfo()?.addObjectGraph(objectGraph)
    }

    @JvmStatic
    fun addLoaderDebug(loaderDebug: LoaderDebug) {
        if (!hasDebugInfo()) {
            return
        }
        getDebugInfo()!!.addLoaderDebug(loaderDebug)
    }

    @JvmStatic
    fun addLoaderDebug(
        entityClass: Class<*>, loaderTimeMillis: Long, ids: Collection<*>, objects: List<*>) {
        if (!hasDebugInfo()) {
            return
        }
        getDebugInfo()?.addLoaderDebug(entityClass, loaderTimeMillis, ids, objects)
    }
}

